### ISSUE TITLE
Beefing Up Security

### ISSUE THREAD

**Issue by thombruce** [thombruce/credible#22]
Came across this Reddit post last night that suggests we should never store JWTs or "authoritative tokens" in a client-readable solution like cookies or localStorage: https://www.reddit.com/r/rails/comments/g7zhjp/beefing_security_of_rails_api/foks4xw/

Currently, Credible... actually doesn't do this, this is a [Helvellyn JS](https://github.com/thombruce/helvellyn.js#helvellynjs) concern. Helvellyn JS receives the JWT and stores it locally, and persists it between sessions.

Our current JWT:

- does not expire
- does not specify its issued at time

We just store it and persistently use it.

The redditor recommends not storing it, but instead using it only in memory... which is fine, we have all manner of ways of handling that from the Vuex store to props and computed properties and whatnot...

They recommend using a Refresh Token in storage to submit to the server and grant a JWT that would then be used in subsequent requests during the same session.

Our JWTs don't even check session tokens, they perform a lookup by ID:

```
session = ::Session.find(token[0]['data']['session_id'])
```

...and they do that on every request... which... they must, because we need the Session in memory to authorize actions, so that's fine.

But this **could** be hijacked.

My question is...

### How would a refresh token protect against this hijacking?

Well...

Consider that we already have Session stored remotely and can invalidate the Session with a `.destroy`. Because the JWT contains an ID, and this value is protected by the JWT signature, if we destroy the Session, we invalidate the JWT.

A refresh token... wouldn't inherently add much security. It could still be read client-side, and hijacked... it would just not contain any authoritative information, like the user's email address. That's a slight improvement.

It would also be invalidatable in the same way JWTs are, by destroying the Session server-side.

It _might_ introduce a vulnerability to [timing attacks](https://github.com/thombruce/credible/issues/2): https://github.com/thombruce/credible/issues/2 This can be avoided by performing a lookup on a separate property. Perhaps a refresh token could be 2-part: **id-token**. Split on '-', lookup ID part and secure compare token part.

Okay, so no timing attack vulnerability if done that way...

More discussion about JWT and refresh token handling can be found here: https://stackoverflow.com/questions/27726066/jwt-refresh-token-flow Including: **Refresh token is just another JWT with fewer claims** In our case, it could contain a session ID for lookup and a "refresh token" for secure compare. Successful validation of which would...

- issue a JWT auth token
- renew the refresh token? This would keep the user's session valid for as long as they continue to visit

What am I missing? How is this not the same as what we have already?

Because:

- **We don't yet timeout or validate date of JWT**
- **User and Session details in the auth JWT would be more secure for not being persisted**

But... if the refresh token can still be hijacked, any potential attacker can still issue a request for those details anyway?? Yup... :\

### Suggestion

1. Minimum security is "good enough" default - this should be as sophisticated as possible while still meeting user expectations like "rememberable"
2. Extended security is configurable - we'll give this more consideration as each "extended" feature is added

Uhmm... So yes, we will set up Credible to support Refresh Tokens because:

- It removes user/session/authorisation information from the auth token

It may not provide additional security from that being accessed in principle, but it does mitigate the problem by requiring a request to access it and:

- The user can revoke sessions remotely (still need to set this up)
- Refresh token will timeout

**Issuing a refreshed refresh token is a vulnerability. The point of them being longlived but still expiring is to limit time in which they can be used by a nefarious party - the user should be required to login when they expire!**

An alternative is to compare usage of a refresh token to revoke it if used "in parallel"; meaning two sessions, each requesting auth tokens from separate clients. This would be indicative of a breach: https://security.stackexchange.com/a/135850

^^^ Looking for such behaviour and revoking based on it is one way in which Refresh Tokens can be _allowed_ to... live forever, or be renewed continuously.

It's an alternative approach though, and many - including Google - favour the revocation of refresh tokens on a half-monthly or monthly basis to mitigate longterm usage.

Ideally... both would be present. And configuration would be open to the user of Credible to decide which of these they want... So, aah... Let's just list the options here for clarity:

1. Refresh token **can** be refreshed indefinitely, but comparison of usage by multiple clients should be watched for and the token revoked if suspicious.
2. Refresh tokens should otherwise be timeoutable and revoked at the end of determined period (typically 14-60 days - this can certainly be set lower if user wishes, higher... increases vulnerability - default will likely be 14 days).

**Comment by thombruce** [comment]
A fairly typical schema presently looks like this:

```ruby
  create_table "sessions", force: :cascade do |t|
    t.bigint "user_id", null: false
    t.string "token"
    t.datetime "created_at", precision: 6, null: false
    t.datetime "updated_at", precision: 6, null: false
    t.index ["token"], name: "index_sessions_on_token", unique: true
    t.index ["user_id"], name: "index_sessions_on_user_id"
  end

  create_table "users", force: :cascade do |t|
    t.string "email"
    t.string "password_digest"
    t.string "confirmation_token"
    t.datetime "confirmed_at"
    t.datetime "created_at", precision: 6, null: false
    t.datetime "updated_at", precision: 6, null: false
    t.index ["confirmation_token"], name: "index_users_on_confirmation_token", unique: true
    t.index ["email"], name: "index_users_on_email", unique: true
  end

  add_foreign_key "sessions", "users"
```

We use a combination of User#email and User#confirmation_token provided as separate values to confirm user email addresses.

We use Session#token to authenticate API access for consuming applications. It currently serves no other purpose and its position here is unclear.

A JWT is issued and a User session is authenticated by decoding the token server-side and looking up the Session#id.

### New Implementation

...okay... so let's split up what needs doing. We want:

- Additional claims such as Issued At Time, Expiration Time to be supported on JWTs (this is the first step towards invalidating out of date tokens)
- We want a distinct type of token - Refresh Token (also a JWT) - to be issued at first authentication and used to renew authentication tokens (priorly referred to as JWT everywhere I've talked about it)
- ...

...and more. But that's a start. We should create two separate issues for those tasks. And we'll have to do some work on https://github.com/thombruce/helvellyn.js and https://github.com/thombruce/credible-demo as well to properly utilise the new format.

**Comment by thombruce** [comment]
closed by #27
